elements/lisk-tree/src/merkle_tree.ts

```diff
@@ -18,62 +18,155 @@ import { hash } from '@liskhq/lisk-cryptography';
 import { Proof } from './types';
 import { EMPTY_HASH, LEAF_PREFIX, BRANCH_PREFIX } from './constants';
 
+const LAYER_INDEX_SIZE = 1;
+const NODE_INDEX_SIZE = 8;
+const NODE_HASH_SIZE = 32;
+
+enum NodeType {
+	ROOT = 'root',
+	BRANCH = 'branch',
+	LEAF = 'leaf',
+}
+
+interface NodeData {
+	readonly value: Buffer;
+	readonly hash: Buffer;
+}
+
+interface NodeInfo {
+	readonly type: NodeType;
+	readonly hash: Buffer;
+	readonly value: Buffer;
+	readonly leftHash: Buffer;
+	readonly rightHash: Buffer;
+	readonly layerIndex: number;
+	readonly nodeIndex: bigint;
+}
+
+const isLeaf = (value: Buffer): boolean =>
+	value[0] === LEAF_PREFIX[0] && value.compare(Buffer.alloc(0)) !== 0;
+
 export class MerkleTree {
 	private _root: Buffer;
-	private _dataLength = 0;
-	private _data: { [key: string]: Buffer } = {};
+	private _width = 0;
+
+	// Object holds data in format { [hash]: value }
+	private _hashToBuffer: { [key: string]: Buffer } = {};
 
 	public constructor(initValues: Buffer[] = []) {
-		this._dataLength = initValues.length;
-		if (this._dataLength === 0) {
+		if (initValues.length === 0) {
```

Comment:
Even for an empty tree, I would insert the emptyHash in `_hashToBuffer`, for consistency